### PR TITLE
integrate tokio-epoll-uring as alternative VirtualFile IO engine

### PR DESCRIPTION
**Pull request by problame**
(part of https://github.com/neondatabase/neon/issues/4744 )

This PR integrates [`tokio-epoll-uring`](https://github.com/neondatabase/tokio-epoll-uring) as an alternative IO engine for Pageserver's `VirtualFile` `open` and `read` operations.

The IO engine is configurable via pageserver config, and for Rust unit tests, via an env var.

The goal of integrating early is to uncover unknown problems with `tokio-epoll-uring` early
by exposing the code to staging & CI tests for a few days. Ideally we'd run all the CI tests for both IO engines, but, that would require twice the CI resources.

On older kernel versions, io_uring SQ and CQ is accounted as locked memory.
This PR raises the memlock ulimit wherever we execute our Rust code in CI.
See #6373 for details on memlock.
The PR that will raise memlock ulimit for staging and prod is neondatabase/aws#932;

### CODE REVIEW

**Comment by problame** [comment]
Should cover the VirtualFile write path as well, i.e., write & fsync.

ref https://github.com/neondatabase/neon/pull/6123#discussion_r1432912611

File: .github/workflows/build_and_test.yml
Comment (problame) [inline .github/workflows/build_and_test.yml:431]: Unless we can expand the CI runner pool, I'll merge this PR with
```
pageserver_virtual_file_io_engine: [ std-fs ]
```

Shortly after the weekly release, I'll then switch it to 

```
pageserver_virtual_file_io_engine: [ tokio-epoll-uring ]
```

for 2-3 days, so that the tokio-epoll-uring code gets coverage.

Then I'll switch it back, so that in the days leading up to the subsequent release, we test with the IO engine we're using in production.

Comment (bayandin) [reply .github/workflows/build_and_test.yml:431]: I want to clean up the python part that I've added a bit.  
I've assigned myself as a reviewer, please don't merge without my approval 🙏

**Review by bayandin — APPROVED** [review]

File: pageserver/src/lib.rs
Comment (problame) [reply pageserver/src/lib.rs:1]: IDK, it fails to compile without it. Compiler suggested to add this.

File: pageserver/src/tenant/block_io.rs
Comment (problame) [reply pageserver/src/tenant/block_io.rs:73]: Hmm, I get

```
diff --git a/pageserver/src/tenant/block_io.rs b/pageserver/src/tenant/block_io.rs
index 8c9294bb9..6035122d5 100644
--- a/pageserver/src/tenant/block_io.rs
+++ b/pageserver/src/tenant/block_io.rs
@@ -68,9 +68,7 @@ impl<'a> Deref for BlockLease<'a> {
             #[cfg(test)]
             BlockLease::Vec(v) => {
                 let v: &Vec<u8> = v;
-                assert_eq!(v.len(), PAGE_SZ, "caller must ensure that v has PAGE_SZ");
-                // Safety: see above assertion.
-                unsafe { &*(v.as_ptr() as *const [u8; PAGE_SZ]) }
+                TryFrom::try_from(v).expect("caller must ensure that v has PAGE_SZ");
             }
         }
     }
```

```
error[E0277]: the trait bound `&[u8; 8192]: std::convert::From<&std::vec::Vec<u8>>` is not satisfied
```

Comment (problame) [reply pageserver/src/tenant/block_io.rs:73]: Ah, I got it worked out

**Review by arpad-m — APPROVED** [review]

File: pageserver/src/virtual_file.rs
Comment (VladLazar) [inline pageserver/src/virtual_file.rs:549]: Does the `buf.is_empty()` check on line 564 still work as intended? IIUC the `slice` call on line 557 just advances the `begin` offset of the slice, so the underlying storage remains the same (empty or not).

File: pageserver/src/virtual_file.rs
Comment (arpad-m) [reply pageserver/src/virtual_file.rs:694]: Not sure about the intent, for which I defer to @problame, but `let _ = foo()` does not keep around the thing returned by `foo()` but drops it immediately (while `let _foobar = foo()` keeps it around). It's a common gotcha.

https://github.com/rust-lang/rust-clippy/issues/8246

**Review by jcsp — APPROVED** [review]
I'm good with this: only doubt is whether any of the VirtualFile changes might introduce subtle perf regressions, but that's what our shiny new benchmarks are for.

**Comment by problame** [comment]
Status update: the Rust unit test failures are due to OOMs. Thanks @koivunej for the #6389 

Example output:

```
2024-01-19T16:09:45.3931975Z --- STDOUT:              pageserver tenant::ephemeral_file::tests::test_ephemeral_blobs ---
2024-01-19T16:09:45.3932930Z◀
2024-01-19T16:09:45.3933073Z running 1 test
2024-01-19T16:09:45.3933620Z test tenant::ephemeral_file::tests::test_ephemeral_blobs ... FAILED
2024-01-19T16:09:45.3934129Z◀
2024-01-19T16:09:45.3934277Z failures:
2024-01-19T16:09:45.3934476Z◀
2024-01-19T16:09:45.3934603Z failures:
2024-01-19T16:09:45.3935026Z     tenant::ephemeral_file::tests::test_ephemeral_blobs
2024-01-19T16:09:45.3935437Z◀
2024-01-19T16:09:45.3935934Z test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 106 filtered out; finished in 0.18s
2024-01-19T16:09:45.3936716Z◀
2024-01-19T16:09:45.3936720Z◀
2024-01-19T16:09:45.3937285Z --- STDERR:              pageserver tenant::ephemeral_file::tests::test_ephemeral_blobs ---
2024-01-19T16:09:45.3939145Z thread 'tenant::ephemeral_file::tests::test_ephemeral_blobs' panicked at .cargo/git/checkouts/tokio-epoll-uring-f670d5b919cd2b86/0dd3a2f/tokio-epoll-uring/src/system/lifecycle.rs:56:82:
2024-01-19T16:09:45.3940590Z called `Result::unwrap()` on an `Err` value: Os { code: 12, kind: OutOfMemory, message: "Cannot allocate memory" }
2024-01-19T16:09:45.3941368Z stack backtrace:
2024-01-19T16:09:45.3941609Z    0: rust_begin_unwind
2024-01-19T16:09:45.3942052Z              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
2024-01-19T16:09:45.3942558Z    1: core::panicking::panic_fmt
2024-01-19T16:09:45.3943030Z              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
2024-01-19T16:09:45.3943531Z    2: core::result::unwrap_failed
2024-01-19T16:09:45.3943999Z              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1653:5
2024-01-19T16:09:45.3944613Z    3: <core::result::Result<io_uring::IoUring, std::io::error::Error>>::unwrap
2024-01-19T16:09:45.3945235Z              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1077:23
2024-01-19T16:09:45.3945881Z    4: <tokio_epoll_uring::system::lifecycle::System>::launch_with_testing::{closure#0}
```

---

I'll investigate that.

Questions to answer:

1. Why does the panic make the test hang?
2. the io_uring ring buffers are mlocked; could the Docker environment limit that to a low number / the high degree of concurrency on the `large-*` runners (32 cores => 32 concurrent nextest) be the limiting factor?

**Comment by problame** [comment]
I hijacked a `large-` runner. Inside the docker container (`docker exec`), here's ulimit output

```
nonroot@1c45f8fd824d:/__w/neon/neon$ ulimit -a
real-time non-blocking time  (microseconds, -R) unlimited
core file size              (blocks, -c) unlimited
data seg size               (kbytes, -d) unlimited
scheduling priority                 (-e) 0
file size                   (blocks, -f) unlimited
pending signals                     (-i) 514764
max locked memory           (kbytes, -l) 64
max memory size             (kbytes, -m) unlimited
open files                          (-n) 1048576
pipe size                (512 bytes, -p) 8
POSIX message queues         (bytes, -q) 819200
real-time priority                  (-r) 0
stack size                  (kbytes, -s) 8192
cpu time                   (seconds, -t) unlimited
max user processes                  (-u) unlimited
virtual memory              (kbytes, -v) unlimited
file locks                          (-x) unlimited
```

=> 64k of locked memory per process.

File: pageserver/src/virtual_file.rs
Comment (problame) [reply pageserver/src/virtual_file.rs:706]: It's intended. The important part is to do the `into_raw_fd()` so the `File::drop` doesn't close it.

RawFd::drop is a no-op.

Not sufficiently clear? How would you write it instead?

Comment (VladLazar) [reply pageserver/src/virtual_file.rs:706]: In hindsight it's clear enough. I raised an eyebrow when reading it for the first time. I guess you could also `mem::forget` but I don't think that's any better.

File: pageserver/src/virtual_file.rs
Comment (problame) [reply pageserver/src/virtual_file.rs:556]: Not sure I understand you concern, let's be thorough here as I found this `Slice` and `BoundedBuf` API to be more tricky than it seems.

- The `self.read_at` calls `buf.bytes_init(n)`, then returns the `buf`
- We then do `buf = buf.slice(n..)`; we REASSIGN local variable `buf` to a new view that is like the old view, but with `n` bytes trimmed at the front, i.e., we're advancing `Slice::front`
- rinse and repeat until `buf.bytes_total() == 0` or EOF

You agree with me up to this point? Please be super nitpicky.

Now, `is_emtpy()` is from the `impl Deref for Slice { Target = [u8] }`.
That `Deref` impl only returns the &[u8] on the initialized portion of the slice.
So yeah, `is_empty()` could indeed return `true` whereas while actually we had an EOF and didn't fill up the buffer fully.

Is that what you were concerned about? I'll come up with a fix.

Comment (problame) [reply pageserver/src/virtual_file.rs:555]: Refactored & pushed fixes. Please review them, reply here, and resolve the conversation if applicable, or add review comments 🙏

Comment (VladLazar) [reply pageserver/src/virtual_file.rs:555]: Right. Agree with your analysis. Went over the code again and it looks good to me. Nice tests!

**Pull request merged**
